### DISCUSSION TITLE
MDAnalysis Universe error on PSF files

### DISCUSSION THREAD

**Discussion by mackevinbraza** [MDAnalysis/mdanalysis#4724] [Q&A] [answered]
Hello all,

I want to ask for help in debugging an MDAnalysis universe error (reading PSF). I have these two PSFs where file QMESA works but PGFSA does not. I attached the two PSF file here. I generated them from cpptraj, where I stripped water, ions, and membrane.  

[zipfile.zip](https://github.com/user-attachments/files/17218432/zipfile.zip)

When I read these PSF in MDanalysis, the following error appeared

```python
ValueError: Failed to construct topology from file /net/gpfs-amarolab/PGFSA.stripped.step5_input.psf with parser <class 'MDAnalysis.topology.PSFParser.PSFParser'>.
Error: Expected section NPHI but found 5
``` 

The PSF files worked well when I load them in VMD along with the trajectories from my MD simulations. 

Not sure on what else can I do to debug this or if there are other things I can do.

Do you know how I can debug this properly? 

Thank you so much!

Sincerely,
Kevin

**Comment by orbeckst** [comment]
In your file `PGFSA.stripped.step5_input.psf` somehow a newline is missing before 8443:
```
    3117    3115    3121    3115    3121    3122    3115    3121    3123
    3122    3121    3123
    8443 !NPHI: dihedrals
       1       5       7       8       2       1       5       6
       2       1       5       7       2       1       5      15
```
When I manually rewrite it to 
```
    3122    3121    3123
    
    8443 !NPHI: dihedrals
       1       5       7       8       2       1       5       6
```
then MDAnalysis can read it.

Will have to look if PSF _must_ have a newline there (then it's a bug in the code you're using) or if MDAnalysis is to strict (then it's a bug in MDAnalysis).

**Comment by yuxuanzhuang** [comment, marked as answer]
Based on the official charmm PSFParser (charmm/source/io/psfres.F90), a blank line is required.
```fortran
     if (qextfmt) then
        idleng=8
        if (prnlev >= 2) write(outu,'(a)') &
             ' psf_read_formatted: Reading PSF in the expanded format.'
        fmt00='(/I10)'
        fmt01=  '(I10,1X,A8,1X,A8,1X,A8,1X,A8,1X,I4,1X,2G14.6,I8)'
        ....
     else
        idleng=4
        fmt00='(/I8)'
        fmt01='(I8,1X,A4,1X,A4,1X,A4,1X,A4,1X,I4,1X,2G14.6,I8)'
        .....
     endif
```
The formats used for both reading and writing have a leading slash, e.g. `fmt00='(/I10)'` or `fmt00='(/I8)'`. In Fortran, the leading `/` in a format skips to the next record. On write, it emits a blank line first; on read, it consumes a blank line before trying to read the integer. If the blank line is missing, the reader will skip the line with the count and then try to read the next line as the integer, causing a mismatch/EOF.

So I think MDAnalysis is handling the parsing correctly here.

**Reply by orbeckst** [reply]
Thank you for checking @yuxuanzhuang . This confirms that it is not a bug in MDAnalysis but in the external program that wrote the other file.
